### ISSUE TITLE
Generating Ghidra/IDA scripts like Il2CppDumper

### ISSUE THREAD

**Issue by ThreeTNT** [vfsfitvnm/frida-il2cpp-bridge#322]
It would help reverse engineering the source code instead of having to guess based on `trace` and `dump.cs`

**Comment by commonuserlol** [comment]
actually il2cppdumper just copy ready2use script. for example you can use script from https://github.com/vfsfitvnm/frida-il2cpp-bridge/discussions/113, but it sets only method name, for fields, strings, etc metadata dump still needed

**Comment by vfsfitvnm** [comment]
Agree, but I doubt I'll work on this anytime soon. I made few attempts in the past, but I always gave up at a certain point.
PS: I won't (personally) support IDA (can't test/use it).

If you want to contribute, this is your chance :smile:

**Comment by vfsfitvnm** [comment]
For instance, the solution would consist in generating a machine readable metadata file (i.e. a json file or a db) so that we can use the target disassembler scripting language to load the metadata into the project.

OT: We could make things even bigger by using the very same metadata to generate some sort of type safe typescript bindings, so that:
```ts
Il2Cpp.corlib.class("System.Int32").new();
```
becomes:
```ts
new mscorlib.System.Int32();
```

**Comment by commonuserlol** [comment]
> Agree, but I doubt I'll work on this anytime soon. I made few attempts in the past, but I always gave up at a certain point. PS: I won't (personally) support IDA (can't test/use it).
> 
> If you want to contribute, this is your chance smile

actually i dont have idea how to find metadata in memory.. im tried [this](https://github.com/350030173/global-metadata_dump/blob/master/global-metadata_dump.js) on 3 apps (android), but no one found metadata... also im tried [this one](https://github.com/vfsfitvnm/frida-il2cpp-bridge/commit/534a18db814543c6184188f75d2fd49e4de62cde) but cannot find `il2cpp_build_metadata` export

**Comment by vfsfitvnm** [comment]
`il2cpp_build_metadata` was a `CModule` function I implemented by my own.
By "metadata" I didn't mean `global-metadata.dat`, but just a machine readable file that contains any metadata of the application we can retrieve (just like `global-metadata.dat`). It's just like `Il2Cpp::dump`, but generate a `json` file instead of a pseudo `cs` file.

**Comment by commonuserlol** [comment]
yes i understood this, but first need to find metadata in memory.. i have one idea and I'll try implement it soon as I can

**Comment by commonuserlol** [comment]
yeah, im generated pseudo metadata .json file, now need figure how to correctly generate .h file and apply c++ signatures..

**Comment by commonuserlol** [comment]
unfortunately I have very little knowledge about the internals of il2cpp so I'll leave it here, maybe it will be useful to someone for further development (this is not the final version)
and yes, the code is not really good
`src/il2cpp/dump.ts`
```
let Metadata = {
                "ScriptMethod": [],
                "ScriptString": [],
                "ScriptMetadata": [],
                "Addresses": [] //not used right now
            };

            for (const assembly of Il2Cpp.domain.assemblies) {
                inform(`dumping ${assembly.name}...`);

                for (const klass of assembly.image.classes) {
                    
                    for (const method of klass.methods) {

                        //@ts-ignore
                        Metadata.ScriptMethod.push({
                            "Address": method.relativeVirtualAddress.toUInt32(),
                            "Name": `${method.class.type.name}$$${method.name}`,
                            "Signature": "soon",
                            "TypeSignature": "soon"
                        });
                    }
                    
                    for (const field of klass.fields) {
                        //@ts-ignore
                        Metadata.ScriptMetadata.push({
                            "Address": klass.handle.sub(Il2Cpp.module.base).add(field.offset).toUInt32(),
                            "Name": field.type.name,
                            "Signature": "soon"
                        });
                    }
                }
            }

            //NOTE: this wont dump ALL strings, only which in memory right now
            inform("dumping strings...");
            for (const str of (Il2Cpp.gc.choose(Il2Cpp.corlib.class("System.String")))) {

                //@ts-ignore
                Metadata.ScriptString.push({
                    "Address": str.handle.sub(Il2Cpp.module.base).toUInt32(),
                    "Value": (new Il2Cpp.String(str)).content ?? ""
                });
            }
            file.write(JSON.stringify(Metadata, null, 4));
            file.flush();
            file.close();
            ok(`dump saved to ${destination}`);
```

**Comment by vfsfitvnm** [comment]
Hi, starting with `frida-il2cpp-bridge` 0.10 (not released yet) we are closer in implemeting https://github.com/vfsfitvnm/frida-il2cpp-bridge/issues/322#issuecomment-1613549080.

As of now, only dumping to C# is possible, but we have the foundations to build other features:
- Generate Ghidra/IDA script so that functions have names and pointer arithmetic in decompiled code is converted to `struct->field` notation;
- Generate `.ts`/`.d.ts` to use TS as it was C#
